README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> MySQL components for Python

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The module contains the following packages:
 
- **Build** - a standard factory for constructing components
- **Connect** - instruments for configuring connections to the database.
- **Persistence** - abstract classes for working with the database that can be used for connecting to collections and performing basic CRUD operations

<a name="links"></a> Quick links:

* [Configuration](https://www.pipservices.org/recipies/configuration)
* [API Reference](https://pip-services3-python.github.io/pip-services3-mysql-python/index.html)
* [Change Log](CHANGELOG.md)
* [Get Help](https://www.pipservices.org/community/help)
* [Contribute](https://www.pipservices.org/community/contribute)

## Use

Install the Python package as
```bash
pip install pip-services3-mysql
```

## Develop

For development you shall install the following prerequisites:
* Python 3.7+
* Visual Studio Code or another IDE of your choice
* Docker

Install dependencies:
```bash
pip install -r requirements.txt
```

Run automated tests:
```bash
python test.py
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized build and test as:
```bash
./build.ps1
./test.ps1
./clear.ps1
```

## Contacts

The library is created and maintained by:
- **Sergey Seroukhov**
- **Danil Prisiazhnyi**

The documentation is written by **Mark Makarychev**.


CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> MySQL components for Python Changelog

## <a name="3.2.4"></a> 3.2.4 (2021-11-25)

### Bug fixes
**persistence** fixed get_one_random

## <a name="3.2.2-3.2.3"></a> 3.2.2-3.2.3 (2021-11-22)

### Bug fixes
* Updated requirements
* Optimize imports
* Fixed connection timeout

## <a name="3.2.1"></a> 3.2.1 (2021-08-25)

### Featuresquit
* Added _request method for PostgresPersistence

## <a name="3.2.0"></a> 3.2.0 (2021-08-09)

Added support for database schemas
t push
### Features
* Added schemas to MySqlPersistence, IdentifiableMySqlPersistence, IdentifiableJsonMySqlPersistence
* Added _auto_generate_id flag to IdentifiableMySqlPersistence

### Bug fixes
* Fixed MySQLConnection open error logging

## <a name="3.1.0"></a> 3.1.0 (2021-05-14)

### Features
* Moved MySQLConnection to **connect** package
* Added type hints
* Fixed returned types for operations

## <a name="3.0.1"></a> 3.0.1 (2021-03-11)

### Bug fixes
* fixed MySqlPersistence.get_page_by_filter sort param
* fixed MySqlPersistence.get_list_by_filter return value

## <a name="3.0.0"></a> 3.0.0 (2021-02-25) 

Initial public release

### Features
* Added DefaultMySqlFactory
* Added MySqlConnectionResolver
* Added IdentifiableJsonMySqlPersistence
* Added IdentifiableMySqlPersistence
* Added MySqlConnection
* Added MySqlPersistence

pip_services3_mysql/connect/MySqlConnectionResolver.py
# -*- coding: utf-8 -*-
from typing import List, Optional

from pip_services3_commons.config import IConfigurable, ConfigParams
from pip_services3_commons.errors import ConfigException
from pip_services3_commons.refer import IReferenceable, IReferences
from pip_services3_components.auth import CredentialResolver, CredentialParams
from pip_services3_components.connect import ConnectionResolver, ConnectionParams


class MySqlConnectionResolver(IReferenceable, IConfigurable):
    """
    Helper class that resolves MySQL connection and credential parameters,
    validates them and generates a connection URI.

    It is able to process multiple connections to MySQL cluster nodes.

    ### Configuration parameters ###
        - connection(s):
            - discovery_key:               (optional) a key to retrieve the connection from :class:`IDiscovery <pip_services3_components.connect.IDiscovery.IDiscovery>`
            - host:                        host name or IP address
            - port:                        port number (default: 27017)
            - database:                    database name
            - uri:                         resource URI or connection string with all parameters in it
        - credential(s):
            - store_key:                   (optional) a key to retrieve the credentials from :class:`ICredentialStore <pip_services3_components.auth.ICredentialStore.ICredentialStore>`
            - username:                    user name
            - password:                    user password

    ### References ###
        - `*:discovery:*:*:1.0`        (optional) :class:`IDiscovery <pip_services3_components.connect.IDiscovery.IDiscovery>` services
        - `*:credential-store:*:*:1.0` (optional) :class:`ICredentialStore <pip_services3_components.auth.ICredentialStore.ICredentialStore>` stores to resolve credentials
    """

    def __init__(self):
        # The connections resolver.
        self._connection_resolver: ConnectionResolver = ConnectionResolver()
        # The credentials resolver.
        self._credential_resolver: CredentialResolver = CredentialResolver()

    def configure(self, config: ConfigParams):
        """
        Configures component by passing configuration parameters.

        :param config: configuration parameters to be set.
        """
        self._connection_resolver.configure(config)
        self._credential_resolver.configure(config)

    def set_references(self, references: IReferences):
        """
        Sets references to dependent components.

        :param references: references to locate the component dependencies.
        """
        self._credential_resolver.set_references(references)
        self._connection_resolver.set_references(references)

    def __validate_connection(self, correlation_id: Optional[str], connection: ConnectionParams):
        uri = connection.get_uri()
        if uri:
            return

        host = connection.get_host()
        if not host:
            raise ConfigException(correlation_id, "NO_HOST", "Connection host is not set")

        port = connection.get_port()
        if port == 0:
            raise ConfigException(correlation_id, "NO_PORT", "Connection port is not set")

        database = connection.get_as_nullable_string('database')
        if not database:
            raise ConfigException(correlation_id, "NO_DATABASE", "Connection database is not set")

    def __validate_connections(self, correlation_id: Optional[str], connections: List[ConnectionParams]):
        if not connections or len(connections) == 0:
            raise ConfigException(correlation_id, "NO_CONNECTION", "Database connection is not set")

        for connection in connections:
            self.__validate_connection(correlation_id, connection)

    def __compose_uri(self, connections: List[ConnectionParams], credential: CredentialParams) -> str:
        # If there is a uri then return it immediately
        for connection in connections:
            uri = connection.get_uri()
            if uri:
                return uri

        # Define hosts
        hosts = ''
        for connection in connections:
            host = connection.get_host()
            port = connection.get_port()

            if len(hosts) > 0:
                hosts += ','
            hosts += host + ('' if port is None else ':' + str(port))

        # Define database
        database = ''
        for connection in connections:
            database = database or connection.get_as_nullable_string('database')

        if len(database) > 0:
            database = '/' + database

        # Define authentication part
        auth = ''
        if credential:
            username = credential.get_username()
            if username:
                password = credential.get_password()
                if password:
                    auth = username + ':' + password + '@'
                else:
                    auth = username + '@'

        # Define additional parameters parameters
        options = ConfigParams.merge_configs(*connections).override(credential)
        options.remove('uri')
        options.remove('host')
        options.remove('port')
        options.remove('database')
        options.remove('username')
        options.remove('password')
        params = ''
        keys = options.get_keys()
        for key in keys:
            if len(params) > 0:
                params += '&'

            params += key

            value = options.get_as_string(key)
            if value:
                params += '=' + value

        if len(params) > 0:
            params = '?' + params

        # Composer uri
        uri = "mysql://" + auth + hosts + database + params

        return uri

    def resolve(self, correlation_id: Optional[str]):
        """
        Resolves MySQL connection URI from connection and credential parameters.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        :return: resolved URI or raise error.
        """

        connections = self._connection_resolver.resolve_all(correlation_id)
        # Validate connections
        self.__validate_connections(correlation_id, connections)

        credential = self._credential_resolver.lookup(correlation_id)
        # Credentials are not validated right now

        return self.__compose_uri(connections, credential)


pip_services3_mysql/persistence/IdentifiableMySqlPersistence.py
# -*- coding: utf-8 -*-
from copy import deepcopy
from typing import Any, Optional, List, TypeVar

from pip_services3_commons.data import IdGenerator, AnyValueMap

from pip_services3_mysql.persistence.MySqlPersistence import MySqlPersistence

T = TypeVar('T')  # Declare type variable


class IdentifiableMySqlPersistence(MySqlPersistence):
    """
    Abstract persistence component that stores data in MySQL
    and implements a number of CRUD operations over data items with unique ids.
    The data items must implement :class:`IIdentifiable <pip_services3_commons.data.IIdentifiable.IIdentifiable>` interface.

    In basic scenarios child classes shall only override :func:`get_page_by_filter <pip_services3_mysql.persistence.IdentifiableJsonMySqlPersistence.get_page_by_filter>`,
    :func:`get_list_by_filter <pip_services3_mysql.persistence.IdentifiableJsonMySqlPersistence.get_list_by_filter>` or :func:`delete_by_filter <pip_services3_mysql.persistence.IdentifiableJsonMySqlPersistence.delete_by_filter>`
    operations with specific filter function.
    All other operations can be used out of the box.

    In complex scenarios child classes can implement additional operations by
    accessing **self._collection** and **self._model** properties.

    ### Configuration parameters ###
        - table:                  (optional) MySQL table name
        - schema:                 (optional) MySQL schema name
        - connection(s):
            - discovery_key:             (optional) a key to retrieve the connection from :class:`IDiscovery <pip_services3_components.connect.IDiscovery.IDiscovery>`
            - host:                      host name or IP address
            - port:                      port number (default: 27017)
            - uri:                       resource URI or connection string with all parameters in it
        - credential(s):
            - store_key:                 (optional) a key to retrieve the credentials from :class:`ICredentialStore <pip_services3_components.auth.ICredentialStore.ICredentialStore>`
            - username:                  (optional) user name
            - password:                  (optional) user password
        - options:
            - connect_timeout:      (optional) number of milliseconds to wait before timing out when connecting a new client (default: 0)
            - idle_timeout:         (optional) number of milliseconds a client must sit idle in the pool and not be checked out (default: 10000)
            - max_pool_size:        (optional) maximum number of clients the pool should contain (default: 10)

    ### References ###
        - `*:logger:*:*:1.0`           (optional) :class:`ILogger <pip_services3_components.log.ILogger.ILogger>` components to pass log messages components to pass log messages
        - `*:discovery:*:*:1.0`        (optional) :class:`IDiscovery <pip_services3_components.connect.IDiscovery.IDiscovery>` services
        - `*:credential-store:*:*:1.0` (optional) :class:`ICredentialStore <pip_services3_components.auth.ICredentialStore.ICredentialStore>` stores to resolve credentials

    Example:

    .. code-block:: python

        class MyMySqlPersistence(IdentifiableMySqlPersistence):
            def __init__(self):
                super(MyMySqlPersistence, self).__init__("mydata", MyDataMySqlSchema())

            def __compose_filter(self, filter):
                filter = filter or FilterParams()
                criteria = []
                name = filter.get_as_nullable_string('name')
                if name:
                    criteria.append({'name': name})
                return {'$and': criteria} if len(criteria) > 0 else None

            def get_page_by_filter(self, correlation_id, filter, paging):
                return super().get_page_by_filter(correlation_id, self.__compose_filter(filter), paging, None, None)

        persistence = MyMySqlPersistence()
        persistence.configure(ConfigParams.from_tuples(
            "host", "localhost",
            "port", 27017
        ))

        persistence.open('123')
        persistence.create('123', {'id': "1", 'name': "ABC"})
        page = persistence.get_page_by_filter('123', FilterParams.from_tuples('name', 'ABC'), None)
        print(page.data) # Result: { id: "1", name: "ABC" }

        persistence.delete_by_id("123", "1")
        # ...
    """

    def __init__(self, table_name: str = None, schema_name: str = None):
        """
        Creates a new instance of the persistence component.

        :param table_name: (optional) a table_name name.
        :param schema_name: (optional) a schema name
        """
        super(IdentifiableMySqlPersistence, self).__init__(table_name, schema_name)

        # Flag to turn on auto generation of object ids.
        self._auto_generate_id: bool = True

    def _convert_from_public_partial(self, value: Any) -> Any:
        """
        Converts the given object from the public partial format.

        :param value: the object to convert from the public partial format.
        :return: the initial object.
        """
        return self._convert_from_public(value)

    def get_list_by_ids(self, correlation_id: Optional[str], ids: List[Any]) -> List[T]:
        """
        Gets a list of data items retrieved by given unique ids.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        :param ids: ids of data items to be retrieved
        :return: data list
        """
        params = self._generate_parameters(ids)
        query = "SELECT * FROM " + self._quoted_table_name() + " WHERE id IN(" + params + ")"
        result = self._client.query(query, ids)
        items = result['items']

        if items is not None:
            self._logger.trace(correlation_id, "Retrieved %d from %s", len(items), self._table_name)

        items = list(map(self._convert_from_public_partial, items))
        return items

    def get_one_by_id(self, correlation_id: Optional[str], id: Any) -> T:
        """
        Gets a data item by its unique id.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        :param id: an id of data item to be retrieved.
        :return: data item
        """
        query = "SELECT * FROM " + self._quoted_table_name() + " WHERE id=%s"
        params = [id]

        result = self._client.query(query, params)
        item = self._convert_to_public(result['items'][0]) if result['items'] and len(result['items']) == 1 else None

        if not item:
            self._logger.trace(correlation_id, "Nothing found from %s with id = %s", self._table_name, id)
        else:
            self._logger.trace(correlation_id, "Retrieved from %s with id = %s", self._table_name, id)

        return item

    def create(self, correlation_id: Optional[str], item: T) -> Optional[T]:
        """
        Creates a data item.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        :param item: an item to be created.
        :return: created item
        """
        if item is None:
            return

        # Assign unique id
        new_item = deepcopy(item)

        if new_item.id is None and self._auto_generate_id:
            new_item = deepcopy(new_item)
            new_item.id = item.id or IdGenerator.next_long()

        return super().create(correlation_id, new_item)

    def set(self, correlation_id: Optional[str], item: T) -> Optional[T]:
        """
        Sets a data item. If the data item exists it updates it,
        otherwise it create a new data item.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        :param item: a item to be set.
        :return: updated item
        """
        if item is None:
            return

        # Assign unique id
        if item.get('id') is None and self._auto_generate_id:
            item = deepcopy(item)
            item['id'] = item['id'] or IdGenerator.next_long()

        row = self._convert_from_public_partial(item)
        columns = self._generate_columns(row)
        params = self._generate_parameters(row)
        set_params = self._generate_set_parameters(row)
        values = self._generate_values(row)
        values += deepcopy(values)
        values.append(item['id'])

        query = "INSERT INTO " + self._quoted_table_name() + " (" + columns + ") VALUES (" + params + ")"
        query += " ON DUPLICATE KEY UPDATE " + set_params
        query += "; SELECT * FROM " + self._quoted_table_name() + " WHERE id=%s"

        result = self._client.query(query, values)

        new_item = self._convert_to_public(result['items'][0]) if result['items'] and len(
            result['items']) == 1 else None

        self._logger.trace(correlation_id, "Set in %s with id = %s", self._quoted_table_name(),
                           item.id)

        return new_item

    def update(self, correlation_id: Optional[str], item: T) -> Optional[T]:
        """
        Updates a data item.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        :param item: an item to be updated.
        :return: updated item
        """
        if item is None:
            return

        row = self._convert_from_public(item)
        params = self._generate_set_parameters(row)
        values = self._generate_values(row)
        values.append(row['id'])
        values.append(row['id'])

        query = "UPDATE " + self._quoted_table_name() + " SET " + params + " WHERE id=%s"
        query += "; SELECT * FROM " + self._quoted_table_name() + " WHERE id=%s"

        result = self._client.query(query, values)

        new_item = self._convert_to_public(result['items'][0]) if result['items'] and len(
            result['items']) == 1 else None

        self._logger.trace(correlation_id, "Updated in %s with id = %s", self._table_name, new_item.id)

        return new_item

    def update_partially(self, correlation_id: Optional[str], id: Any, data: AnyValueMap) -> Optional[T]:
        """
        Updates only few selected fields in a data item.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        :param id: an id of data item to be updated.
        :param data: a map with fields to be updated.
        :return:  updated item
        """
        if data is None or id is None:
            return

        row = self._convert_from_public_partial(data)
        params = self._generate_set_parameters(row)
        values = self._generate_values(row)
        values.append(id)
        values.append(id)

        query = "UPDATE " + self._quoted_table_name() + " SET " + params + " WHERE id=%s"
        query += "; SELECT * FROM " + self._quoted_table_name() + " WHERE id=%s"

        result = self._client.query(query, values)

        self._logger.trace(correlation_id, "Updated partially in %s with id = %s", self._table_name, id)
        new_item = self._convert_to_public(result['items'][0]) if result['items'] and len(
            result['items']) == 1 else None

        return new_item

    def delete_by_id(self, correlation_id: Optional[str], id: Any) -> T:
        """
        Deleted a data item by it's unique id.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        :param id: an id of the item to be deleted
        :return: deleted item
        """
        values = [id, id]

        query = "SELECT * FROM " + self._quoted_table_name() + " WHERE id=%s"
        query += "; DELETE FROM " + self._quoted_table_name() + " WHERE id=%s"

        result = self._client.query(query, values)

        self._logger.trace(correlation_id, "Deleted from %s with id = %s", self._table_name, id)
        deleted_item = self._convert_to_public(result['items'][0]) if result['items'] and len(
            result['items']) == 1 else None

        return deleted_item

    def delete_by_ids(self, correlation_id: Optional[str], ids: List[Any]):
        """
        Deletes multiple data items by their unique ids.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        :param ids: ids of data items to be deleted.
        :return: None for success
        """
        params = self._generate_parameters(ids)
        query = "DELETE FROM " + self._quoted_table_name() + " WHERE id IN(" + params + ")"

        result = self._client.query(query, ids)
        count = result['rowcount'] if result['rowcount'] else 0
        self._logger.trace(correlation_id, "Deleted %d items from %s", count, self._table_name)


pip_services3_mysql/connect/MySqlConnection.py
# -*- coding: utf-8 -*-

import urllib.parse as urlparse
from typing import Any, Union, Optional

import mysql.connector
from mysql.connector import pooling
from pip_services3_commons.config import IConfigurable, ConfigParams
from pip_services3_commons.errors import ConnectionException
from pip_services3_commons.refer import IReferenceable, IReferences
from pip_services3_commons.run import IOpenable
from pip_services3_components.log import CompositeLogger

from pip_services3_mysql.connect.MySqlConnectionResolver import MySqlConnectionResolver


class MySqlConnection(IReferenceable, IConfigurable, IOpenable):
    """
    MySQL connection using plain driver.

    By defining a connection and sharing it through multiple persistence components
    you can reduce number of used database connections.

    ### Configuration parameters ###
        - connection(s):
            - discovery_key:             (optional) a key to retrieve the connection from :class:`IDiscovery <pip_services3_components.connect.IDiscovery.IDiscovery>`
            - host:                      host name or IP address
            - port:                      port number (default: 27017)
            - uri:                       resource URI or connection string with all parameters in it
        - credential(s):
            - store_key:                 (optional) a key to retrieve the credentials from :class:`ICredentialStore <pip_services3_components.auth.ICredentialStore.ICredentialStore>`
            - username:                  user name
            - password:                  user password
        - options:
            - connect_timeout:      (optional) number of milliseconds to wait before timing out when connecting a new client (default: 0)
            - idle_timeout:         (optional) number of milliseconds a client must sit idle in the pool and not be checked out (default: 10000)
            - max_pool_size:        (optional) maximum number of clients the pool should contain (default: 10)

    ### References ###
        - `*:logger:*:*:1.0`           (optional) :class:`ILogger <pip_services3_components.log.ILogger.ILogger>` components to pass log messages components to pass log messages
        - `*:discovery:*:*:1.0`        (optional) :class:`IDiscovery <pip_services3_components.connect.IDiscovery.IDiscovery>` services
        - `*:credential-store:*:*:1.0` (optional) :class:`ICredentialStore <pip_services3_components.auth.ICredentialStore.ICredentialStore>` stores to resolve credentials
    """

    def __init__(self):
        self.__default_config = ConfigParams.from_tuples(
            "options.connect_timeout", 0,
            "options.idle_timeout", 10000,
            "options.max_pool_size", 3
        )

        # The logger.
        self._logger: CompositeLogger = CompositeLogger()
        # The connection resolver.
        self._connection_resolver: MySqlConnectionResolver = MySqlConnectionResolver()
        # The configuration options.
        self._options: ConfigParams = ConfigParams()
        # The MySQL connection pool object.
        self._connection: Any = None
        # The MySQL database name.
        self._database_name: str = None

    def configure(self, config: ConfigParams):
        """
        Configures component by passing configuration parameters.

        :param config: configuration parameters to be set.
        """
        config = config.set_defaults(self.__default_config)

        self._connection_resolver.configure(config)

        self._options = self._options.override(config.get_section('options'))

    def set_references(self, references: IReferences):
        """
        Sets references to dependent components.

        :param references: references to locate the component dependencies.
        """
        self._logger.set_references(references)
        self._connection_resolver.set_references(references)

    def is_open(self) -> bool:
        """
        Checks if the component is opened.

        :return: true if the component has been opened and false otherwise.
        """
        return self._connection is not None

    def __compose_uri_settings(self, uri: str, return_uri=False) -> Union[dict, str]:
        max_pool_size = self._options.get_as_nullable_integer('max_pool_size')
        connection_timeout_ms = self._options.get_as_nullable_integer('connect_timeout')

        settings = {
            'pool_size': max_pool_size,
            'connection_timeout': connection_timeout_ms / 1000 if connection_timeout_ms > 0 else 5,
        }
        if not return_uri:
            parsed_url = urlparse.urlparse(uri)
            settings.update({'database': parsed_url.path[1:],
                             'user': parsed_url.username,
                             'password': parsed_url.password,
                             'host': parsed_url.hostname,
                             'port': parsed_url.port,
                             })
            return settings

        params = ''
        for key in settings:
            if len(params) > 0:
                params += '&'

            params += key

            value = settings.get(key)
            if value is not None:
                params += '=' + str(value)

        if uri.find('?') < 0:
            uri += '?' + params
        else:
            uri += '&' + params

        return uri

    def open(self, correlation_id: Optional[str]):
        """
        Opens the component.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        """
        try:
            uri = self._connection_resolver.resolve(correlation_id)
            self._logger.debug(correlation_id, "Connecting to mysql")

            try:
                config = self.__compose_uri_settings(uri)
                pool = mysql.connector.pooling.MySQLConnectionPool(pool_size=config.pop('pool_size'),
                                                                   **config)
                # Try to connect
                connection_obj = pool.get_connection()

                # set timeout
                idle_timeout_ms = self._options.get_as_nullable_integer('idle_timeout')
                if idle_timeout_ms:
                    cursor = connection_obj.cursor()
                    cursor.execute(f"SET SESSION MAX_EXECUTION_TIME={idle_timeout_ms}")
                    connection_obj.commit()
                    cursor.close()
                    connection_obj.close()

                if connection_obj:
                    self._connection = pool
                    self._database_name = pool.pool_name.split('_')[-1]  # TODO need get dbname
                    connection_obj.close()
            except Exception as err:
                raise ConnectionException(correlation_id, "CONNECT_FAILED", "Connection to mysql failed").with_cause(
                    err)

        except Exception as err:
            self._logger.error(correlation_id, err, 'Failed to resolve MySql connection')

    def close(self, correlation_id: Optional[str]):
        """
        Closes component and frees used resources.

        :param correlation_id: (optional) transaction id to trace execution through call chain.
        :return: raise error or None no errors occured.
        """
        if self._connection is None:
            return
        try:
            self._connection._remove_connections()
            self._connection = None
            self._logger.debug(correlation_id, "Disconnected from mysql database %s", self._database_name)
        except Exception as err:
            ConnectionException(correlation_id, 'DISCONNECT_FAILED', 'Disconnect from mysql failed: ').with_cause(err)

        self._connection = None
        self._database_name = None

    def get_connection(self) -> Any:
        return self._connection

    def get_database_name(self) -> str:
        return self._database_name
